activesupport/lib/active_support/callbacks.rb

```diff
@@ -621,32 +622,33 @@ def index(o);     @chain.index(o); end
         def empty?;       @chain.empty?; end
 
         def insert(index, o)
-          @callbacks = nil
+          @callbacks.clear
           @chain.insert(index, o)
         end
 
         def delete(o)
-          @callbacks = nil
+          @callbacks.clear
           @chain.delete(o)
         end
 
         def clear
-          @callbacks = nil
+          @callbacks.clear
           @chain.clear
           self
         end
 
         def initialize_copy(other)
-          @callbacks = nil
+          @callbacks = {}
           @chain     = other.chain.dup
           @mutex     = Mutex.new
         end
 
-        def compile
-          @callbacks || @mutex.synchronize do
+        def compile(types)
+          key = types.sort.join("_")
```

Comment:
This sort should not be a big deal, but since this is callbacks, we need to make sure things didn't get slower. For the most common case. Can you benchmark to see if regular callback executing didn't get slower after this PR and share the results?

Reply:
I'd definitely remove the `join` though.

Reply:
<details>
<summary> Benchmark Script </summary>
<pre>
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "benchmark-ips"
end

require "active_support"
require "active_support/core_ext/object/blank"

# Your patch goes here.
module ActiveSupport
  module Callbacks
    def run_callbacks_new(kind, types = CALLBACK_FILTER_TYPES)
      callbacks = __callbacks[kind.to_sym]

      if callbacks.empty?
        yield if block_given?
      else
        env = Filters::Environment.new(self, false, nil)

        next_sequence = callbacks.compile_new(types)

        # Common case: no 'around' callbacks defined
        if next_sequence.final?
          next_sequence.invoke_before(env)
          env.value = !env.halted && (!block_given? || yield)
          next_sequence.invoke_after(env)
          env.value
        else
          invoke_sequence = Proc.new do
            skipped = nil

            while true
              current = next_sequence
              current.invoke_before(env)
              if current.final?
                env.value = !env.halted && (!block_given? || yield)
              elsif current.skip?(env)
                (skipped ||= []) << current
                next_sequence = next_sequence.nested
                next
              else
                next_sequence = next_sequence.nested
                begin
                  target, block, method, *arguments = current.expand_call_template(env, invoke_sequence)
                  target.send(method, *arguments, &block)
                ensure
                  next_sequence = current
                end
              end
              current.invoke_after(env)
              skipped.pop.invoke_after(env) while skipped&.first
              break env.value
            end
          end

          invoke_sequence.call
        end
      end
    end

    class CallbackChain # :nodoc:
      include Enumerable

      attr_reader :name, :config

      def initialize(name, config)
        @name = name
        @config = {
          scope: [:kind],
          terminator: default_terminator
        }.merge!(config)
        @chain = []
        @callbacks_new = {}
        @callbacks = nil
        @mutex = Mutex.new
      end

      def each(&block); @chain.each(&block); end
      def index(o);     @chain.index(o); end
      def empty?;       @chain.empty?; end

      def insert(index, o)
        @callbacks_new.clear
        @callbacks = nil
        @chain.insert(index, o)
      end

      def delete(o)
        @callbacks_new.clear
        @callbacks = nil
        @chain.delete(o)
      end

      def clear
        @callbacks_new.clear
        @callbacks = nil
        @chain.clear
        self
      end

      def initialize_copy(other)
        @callbacks_new = {}
        @callbacks = nil
        @chain     = other.chain.dup
        @mutex     = Mutex.new
      end

      def compile_new(types)
        key = types.sort.join("_")
        @callbacks_new[key] || @mutex.synchronize do
          final_sequence = CallbackSequence.new
          @callbacks_new[key] ||= @chain.reverse.inject(final_sequence) do |callback_sequence, callback|
            types.include?(callback.kind) ? callback.apply(callback_sequence) : callback_sequence
          end
        end
      end

      def append(*callbacks)
        callbacks.each { |c| append_one(c) }
      end

      def prepend(*callbacks)
        callbacks.each { |c| prepend_one(c) }
      end

      protected
        attr_reader :chain

      private
        def append_one(callback)
          @callbacks_new.clear
          @callbacks = nil
          remove_duplicates(callback)
          @chain.push(callback)
        end

        def prepend_one(callback)
          @callbacks_new.clear
          @callbacks = nil
          remove_duplicates(callback)
          @chain.unshift(callback)
        end

        def remove_duplicates(callback)
          @callbacks_new.clear
          @callbacks = nil
          @chain.delete_if { |c| callback.duplicates?(c) }
        end

        def default_terminator
          Proc.new do |target, result_lambda|
            terminate = true
            catch(:abort) do
              result_lambda.call
              terminate = false
            end
            terminate
          end
        end
    end
  end
end

class Record
  include ActiveSupport::Callbacks
  define_callbacks :save

  def save
    run_callbacks :save do
      true
    end
  end
end

class OnlyBefore < Record
  set_callback :save, :before, :before_callback
  set_callback :save, :before do |obj|
    true
  end
  set_callback :save, :before, :another_before

  def before_callback
    true
  end

  def another_before
    true
  end
end

class OnlyAfter < Record
  set_callback :save, :after, :after_callback
  set_callback :save, :after do |obj|
    true
  end
  set_callback :save, :after, :another_after

  def after_callback
    true
  end

  def another_after
    true
  end
end

class BothTypes < Record
  set_callback :save, :before, :before_callback
  set_callback :save, :before do |obj|
    true
  end
  set_callback :save, :before, :another_before

  set_callback :save, :after, :after_callback
  set_callback :save, :after do |obj|
    true
  end
  set_callback :save, :after, :another_after

  def before_callback
    true
  end

  def another_before
    true
  end

  def after_callback
    true
  end

  def another_after
    true
  end
end

class AllTypes < Record
  set_callback :save, :before, :before_callback
  set_callback :save, :before do |obj|
    true
  end
  set_callback :save, :before, :another_before

  set_callback :save, :after, :after_callback
  set_callback :save, :after do |obj|
    true
  end
  set_callback :save, :after, :another_after
  set_callback :save, :around do
    true
  end

  def before_callback
    true
  end

  def another_before
    true
  end

  def after_callback
    true
  end

  def another_after
    true
  end
end

# Enumerate some representative scenarios here.
#
# It is very easy to make an optimization that improves performance for a
# specific scenario you care about but regresses on other common cases.
# Therefore, you should test your change against a list of representative
# scenarios. Ideally, they should be based on real-world scenarios extracted
# from production applications.
SCENARIOS = {
  "OnlyBefore" => OnlyBefore,
  "OnlyAfter" => OnlyAfter,
  "BothTypes" => BothTypes,
  "AllTypes" => AllTypes,
}

SCENARIOS.each_pair do |name, value|
  puts
  puts " #{name} ".center(80, "=")
  puts

  record = value.new

  Benchmark.ips do |x|
    x.report("run_callbacks_old") { 1000.times { record.run_callbacks :save } }
    x.report("run_callbacks_new") { 1000.times { record.run_callbacks_new :save, [:before, :after, :around] } }
    x.compare!
  end
end
</pre>
</details>

```
Fetching https://github.com/rails/rails.git
Fetching gem metadata from https://rubygems.org/.......
Resolving dependencies...
Using rake 13.0.6
Using builder 3.2.4
Using erubi 1.11.0
Using racc 1.6.0
Using crass 1.0.6
Using rack 2.2.4
Using nio4r 2.5.8
Using websocket-extensions 0.1.5
Using zeitwerk 2.6.0
Using marcel 1.0.2
Using nokogiri 1.13.8 (x86_64-linux)
Using rack-test 2.0.2
Using websocket-driver 0.7.5
Using loofah 2.18.0
Using connection_pool 2.2.5
Using rails-html-sanitizer 1.4.3
Using mini_mime 1.1.2
Using timeout 0.3.0
Using strscan 3.0.4
Using thor 1.2.1
Using digest 3.1.0
Using benchmark-ips 2.10.0
Using bundler 2.3.17
Using concurrent-ruby 1.1.10
Using minitest 5.16.3
Using tzinfo 2.0.5
Using net-protocol 0.1.3
Using mail 2.7.1
Using net-pop 0.1.1
Using i18n 1.12.0
Using net-imap 0.2.3
Using activesupport 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using net-smtp 0.3.1
Using rails-dom-testing 2.0.3
Using activemodel 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using globalid 1.0.0
Using actionview 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using activejob 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using activerecord 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using actionpack 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using actioncable 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using activestorage 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using actionmailer 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using railties 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using actionmailbox 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using actiontext 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)
Using rails 7.1.0.alpha from https://github.com/rails/rails.git (at main@01d345c)

================================== OnlyBefore ==================================

Warming up --------------------------------------
   run_callbacks_old    34.000  i/100ms
   run_callbacks_new    21.000  i/100ms
Calculating -------------------------------------
   run_callbacks_old    337.683  (± 7.7%) i/s -      1.700k in   5.078624s
   run_callbacks_new    207.406  (± 2.9%) i/s -      1.050k in   5.066953s

Comparison:
   run_callbacks_old:      337.7 i/s
   run_callbacks_new:      207.4 i/s - 1.63x  (± 0.00) slower


================================== OnlyAfter ===================================

Warming up --------------------------------------
   run_callbacks_old    76.000  i/100ms
   run_callbacks_new    32.000  i/100ms
Calculating -------------------------------------
   run_callbacks_old    769.548  (± 4.9%) i/s -      3.876k in   5.051347s
   run_callbacks_new    333.308  (± 2.4%) i/s -      1.696k in   5.091512s

Comparison:
   run_callbacks_old:      769.5 i/s
   run_callbacks_new:      333.3 i/s - 2.31x  (± 0.00) slower


================================== BothTypes ===================================

Warming up --------------------------------------
   run_callbacks_old    27.000  i/100ms
   run_callbacks_new    17.000  i/100ms
Calculating -------------------------------------
   run_callbacks_old    271.660  (± 5.5%) i/s -      1.377k in   5.087558s
   run_callbacks_new    183.228  (± 3.8%) i/s -    918.000  in   5.017141s

Comparison:
   run_callbacks_old:      271.7 i/s
   run_callbacks_new:      183.2 i/s - 1.48x  (± 0.00) slower


=================================== AllTypes ===================================

Warming up --------------------------------------
   run_callbacks_old    22.000  i/100ms
   run_callbacks_new    15.000  i/100ms
Calculating -------------------------------------
   run_callbacks_old    216.552  (± 6.0%) i/s -      1.100k in   5.099592s
   run_callbacks_new    157.740  (± 2.5%) i/s -    795.000  in   5.043234s

Comparison:
   run_callbacks_old:      216.6 i/s
   run_callbacks_new:      157.7 i/s - 1.37x  (± 0.00) slower
```

Hmmm, I'm not sure why we are having the performance hit since after the first iteration everything is cached so we don't have any real difference from the old other than the hash lookup. I'm not too familiar with benchmarking but is this acceptable or are there any things that stand out that I'm missing that are causing the slowdown?

Reply:
> I'd definitely remove the `join` though.

Perhaps it is just me but something feels wrong about using an array (or if I go this approach could probably just go with a set) as a key in a hash. I don't feel strongly enough though about this that I'm fine with using array or set if that is the direction that is preferred.

Reply:
> I'm not sure why we are having the performance hit since after the first iteration everything is cached so we don't have any real difference from the old other than the hash lookup

Only profiling will tell for sure, but you added:

  - Array#sort
  - Array#join on an array containing symbols, so 3 `Symbol#to_s` as well.
  - A hash lookup with a string, so the string need to be hashes.

So that's 5 object allocations on the fast path, plus the hash lookup, I'm not that surprised it's slower.

Reply:
>It means if types.nil?means all callbacks, then we only add a nil check on the happy path, so the overhead should be super minimal.

Ya that should be good. Only question is what happens if `[:before, :around, :after]` is specified because then it is the same as nil but would go down the hash path. Defaulting to `[:before, :around, :after]` and doing a count check is slower than nil check but would then cover that case. Or we can just let it be a and leave it to the consumer to figure out there is a slight perf hit if they specify all callback types

Reply:
I'm just following this discussion as an outsider, so I apologize if this has already been brought up elsewhere.

I think the example with `insert_all` / `create_all` is very interesting.  It seems like passing a scalar `:before` or `:after` could work equally well, though.  Are there other examples where it is necessary to specify an array?  And could they use multiple `run_callbacks` calls (one per type) instead?  Specifying a symbol or nil (or `:all`) instead of an array might simplify things and eliminate some of the expensive operations.

Reply:
@jonathanhefner to be fair I need to look at the use case more, because I'm not sure how you could run the around callback on many objects with such API.

Because `around` takes a block, you'd need to nest the calls, so if you have say 1k objects, you'd end up with a 1k frames and likely get a StackLevelTooDeep.

So I think it's a question that if it was answered first would allow to better design the API. because indeed if you only allow a single symbol all this code would be much easier to optimize.

Reply:
> It seems like passing a scalar :before or :after could work equally well, though.

For the uses in the `insert_all` case, you are correct that we would only need to run 1 of the types at a time so an array isn't needed. I can't actually think of a case where you would need to run 2 types in 1 execution of `run_callbacks` but I figured building out for that case just in case someone wanted it in the future would be good to do but looking back on it now I think going with the scalar parameter might be better for now and we can always enhance this API again should the use case ever present itself.

>Because around takes a block, you'd need to nest the calls, so if you have say 1k objects, you'd end up with a 1k frames and likely get a StackLevelTooDeep.

This was a concern of mine too, but for our case we aren't expecting to have crazy numbers of objects but either way I was going to be batching our inserts so we would only have a max of maybe 100 per batch. This was again a situation of planning to put the functionality there and let the consumer decide how they want to utilize it.